### PR TITLE
Add event support for the Catalog Plugin GitLab Discovery Entity providers

### PR DESCRIPTION
**Pull request by elaine-mattos**
## Hey, I just made a Pull Request!

Add events support for the GitLab Entity Provider modules. 
Refs: [RFC: Event Subscription Mechanism for Cross-Plugin Module Integration in the New Backend System #22393](https://github.com/backstage/backstage/issues/22393) and [🚀 Feature: Add Events Support for GitlabDiscoveryEntityProvider and GitlabOrgDiscoveryEntityProvider #21553](https://github.com/backstage/backstage/issues/21553)
#### :heavy_check_mark: TODO List

- [x] Add GitLabOrgDiscoveryEntityProvider module (new Backend)
- [x] Add events support for the GitLabDiscoveryEntityProvider
- [x] Add events support for the GitLabOrgDiscoveryEntityProvider
- [x] Add tests
- [x] Update documentation
- [x] Remove EventSubscriber interface implementation (https://github.com/backstage/backstage/pull/17565#discussion_r1184937825)
- [x] Remove the Events Extension Point and add the [Events Service](https://github.com/backstage/backstage/pull/22344)

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts
Comment (vinzscam) [inline plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts:421]: same comment as `all_available: true` but for `INHERITED` 😄

File: .changeset/dull-sloths-clap.md
Comment (vinzscam) [inline .changeset/dull-sloths-clap.md:2]: both look like a `patch` as there will be no breaking change for the adopters, right?

I would suggest splitting this changeset into two separate changesets for each of the packages so that the CHANGELOG for each package will contain only the appropriate changes.

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts
Comment (vinzscam) [inline plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts:362]: Will `all_available: true` change how the provider behaves for existing adopters? If so, perhaps we could make this change optional through some configuration?

**Review by vinzscam — CHANGES_REQUESTED** [review]
amazing work 👏
Just added a tiny comment about the changesets, otherwise ✨

File: plugins/catalog-backend-module-gitlab/src/lib/client.ts
Comment (vinzscam) [inline plugins/catalog-backend-module-gitlab/src/lib/client.ts:394]:
```suggestion
        request.searchParams.append(key, options[key].toString());
```

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts
Comment (elaine-mattos) [reply plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts:362]: Hi @vinzscam ,

so, there is a nice discussion going on about it [here](https://github.com/backstage/backstage/pull/23287). My colleague @markussiebert sent a PR that would allow the parameters for this [api call](https://docs.gitlab.com/ee/api/groups.html#list-groups) to be configurable. However, as @freben said, it would be probably a burden to the majority of the users to have this many options... 

At this time I cannot think of a better solution...I believe having this hardcoded would be benefitial to all users right now as it would allow them to list every group their token has access to, no matter if the token is linked to a regular user or admin. What do you think?

File: plugins/catalog-backend-module-gitlab/src/lib/client.ts
Comment (elaine-mattos) [reply plugins/catalog-backend-module-gitlab/src/lib/client.ts:394]: I would leave this one as it is because we're already checking if the options exist in line 393 ( and line 358, where the same construction was used before). If I remove the non-null assertion operator, even if the check has been done before, Typescript warns me that the "Object is possibly 'undefined'". 
Is there any other way around this?

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts
Comment (markussiebert) [reply plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts:362]: I still think every option that allows filter the results at gitlab api level isn't a bad option ;-)

File: .changeset/dull-sloths-clap.md
Comment (elaine-mattos) [reply .changeset/dull-sloths-clap.md]: Done! :)

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts
Comment (elaine-mattos) [reply plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts]: Done, please let me know your thoughts! :)

File: plugins/catalog-backend-module-gitlab/src/lib/client.ts
Comment (vinzscam) [reply plugins/catalog-backend-module-gitlab/src/lib/client.ts:394]: ah good point I think something like this should work:

```js
    for (const key in options) {
      const value = options[key];
      if (value !== undefined && value !== '') {
        request.searchParams.append(key, value.toString());
      }
    }
```

so that we are more protected in case the `if` gets removed in the future

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts
Comment (vinzscam) [reply plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts:362]: thank you both, I've missed the discussion. Sorry for causing some confusion 😂
Yeah if `all_available: true` is going to "always be set" then I agree, let's keep it hardcoded.

> I still think every option that allows filter the results at gitlab api level isn't a bad option ;-)

that's absolutely true, I agree! But sometimes it's nice to have less code to maintain, so that we only maintain the configurations that matter.

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts
Comment (vinzscam) [inline plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts:749]: if we receive a `user_remove_from_group` event and `getGroupMembers()` returns empty array we don't modify spec.members. Is it correct?

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts
Comment (vinzscam) [inline plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts:731]: if I've read the logic correctly this could be simplified a bit:

```suggestion
    const relations = this.config.allowInherited
      ? ['DIRECT', 'INHERITED']
      : ['DIRECT'];

    const groupMembers = await client.getGroupMembers(
      groupToRebuild.full_path,
      relations,
    );

    // new members of the group
    const usersToBeAdded =
      event.event_name === 'user_remove_from_group'
        ? groupMembers.items.filter(e => e.username !== event.user_username)
        : groupMembers.items;
    }
```

then below
```js
usersToBeAdded.map(e => e.username)
```

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts
Comment (vinzscam) [inline plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts:422]:
```suggestion
        const relations = this.config.allowInherited
          ? ['DIRECT', 'INHERITED']
          : ['DIRECT'];
```

**Review by vinzscam — CHANGES_REQUESTED** [review]

**Comment by elaine-mattos** [comment]
Hi @vinzscam !

I had some issues with my branch and had to recreate it (I know, I'm such a git newbie! :D ).
I did incorporate your suggestions/remarks. Could you please take a look at them?
Also, the pipeline is throwing a formatting error on a file I did not touch (beps/0004-scaffolder-task-idempotency/README.md). Should I do anything about it?

thanks again in advance!!! :)

File: .changeset/eleven-pandas-divide.md
Comment (elaine-mattos) [reply .changeset/eleven-pandas-divide.md]: Done!

File: plugins/catalog-backend-module-gitlab/src/lib/client.ts
Comment (elaine-mattos) [reply plugins/catalog-backend-module-gitlab/src/lib/client.ts]: Done!

File: .changeset/shy-students-clap.md
Comment (elaine-mattos) [reply .changeset/shy-students-clap.md]: Also done!

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts
Comment (elaine-mattos) [reply plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts]: Great remark, thanks! :)

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts
Comment (elaine-mattos) [reply plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts]: Done

**Comment by freben** [comment]
Yeah that `prettier` error has been fixed since then. Rebasing on a fresh master gets rid of the problem. Something like this

```
git checkout master
git pull
git checkout feat/add-event-support-gitlab-provider
git pull
git rebase master
```

**Comment by elaine-mattos** [comment]
Hi guys,

the suggested changes have been made. Could you please take a look?
I appreciate the feedback :)

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts
Comment (elaine-mattos) [reply plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts:730]: Hi, yes, this is correct!

File: plugins/catalog-backend-module-gitlab-org/src/catalogModuleGitlabOrgDiscoveryEntityProvider.ts
Comment (pjungermann) [inline plugins/catalog-backend-module-gitlab-org/src/catalogModuleGitlabOrgDiscoveryEntityProvider.ts:22]: _a question for the maintainers; also rather for a follow-up_

I know that this follows the same pattern as used for the GitHub org provider. However, I have the same question there.

As we create a separate module and package for `gitlab-org`, wouldn't it be better to move the provider to this package and deprecate it at the other one (e.g., by importing it from this new package and reexporting it, marked as deprecated -- similar to few other places).

What do you think?

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts
Comment (elaine-mattos) [reply plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts]: You're right! I have changed it to `const usersToBeAdded = groupMembers.items;`

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts
Comment (elaine-mattos) [reply plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts]: Done! :)

**Comment by elaine-mattos** [comment]
Hi!
I have incorporated the changes requested by @vinzscam .  However, I'm trying to update the feature branch but I'm getting tons of vale errors. My vale version is the 3.3.1. 
I can fix the errors by adding backticks around the unrecognized words in the changelog.md files. The question is, should I? :)

File: plugins/catalog-backend-module-gitlab-org/src/catalogModuleGitlabOrgDiscoveryEntityProvider.ts
Comment (elaine-mattos) [reply plugins/catalog-backend-module-gitlab-org/src/catalogModuleGitlabOrgDiscoveryEntityProvider.ts:22]: Hi @pjungermann ,

 I would be glad to do it if provided with an example! :)

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts
Comment (elaine-mattos) [reply plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts]: Hi, 
yeah, you're right. Again! :)
Is it better now? 

```typescript
    // update members of the group entity
    groupEntityToModify[0].spec.members =
      usersToBeAdded.length !== 0 ? usersToBeAdded.map(e => e.username) : [];
```

Comment (vinzscam) [reply plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts]: yeah sounds good! You could also skip the check `groupEntityToModify[0].spec.members = usersToBeAdded.map(e => e.username);` but it's ok!

**Comment by elaine-mattos** [comment]
Hi @vinzscam ,

great news! 😄 
We caught a small bug in our environment and fixed it. The `allowInherited` config option was not being correctly read from the config file because it was not added to the `readGitlabConfig` function (my bad 🥲 ). I integrated the fix in the last commit and also rebased to get rid of the merge conflict.

File: plugins/catalog-backend-module-gitlab-org/src/catalogModuleGitlabOrgDiscoveryEntityProvider.ts
Comment (pjungermann) [reply plugins/catalog-backend-module-gitlab-org/src/catalogModuleGitlabOrgDiscoveryEntityProvider.ts:22]: @vinzscam yeah, nothing that blocks this PR.

I'm also not sure what the maintainers opinion is on that one. I think it would make sense to split that up.
It does affect other plugins like the one for github-org as well.

Comment (vinzscam) [reply plugins/catalog-backend-module-gitlab-org/src/catalogModuleGitlabOrgDiscoveryEntityProvider.ts:22]: I think the issue is that `gitlab` can't import from `gitlab-org`, as this will cause a circular dependency

**Review by vinzscam — APPROVED** [review]

**Review by awanlin — CHANGES_REQUESTED** [review]
Hi @elaine-mattos, can you please add details about the fact that you need to define the schedule in your provider config when using the new backend system? This is missing in the current docs and I don't see it here. Thanks!

**Comment by elaine-mattos** [comment]
> Hi @elaine-mattos, can you please add details about the fact that you need to define the schedule in your provider config when using the new backend system? This is missing in the current docs and I don't see it here. Thanks!

Hi @awanlin ! I've changed the docu a little bit. Could you please review? :)

File: docs/integrations/gitlab/org.md
Comment (awanlin) [inline docs/integrations/gitlab/org.md:173]: Is this true in all cases that it's optional? If I'm using the New Backend System would that not then be required?

File: docs/integrations/gitlab/discovery.md
Comment (awanlin) [inline docs/integrations/gitlab/discovery.md:153]: Is this true in all cases that it's optional? If I'm using the New Backend System would that not then be required?

**Review by awanlin — COMMENTED** [review]
Thanks for adding in the schedule details to the docs @elaine-mattos, I left two comments but they are the same. As I haven't reviewed the code just the docs I'm only leaving this review as a comment

File: docs/integrations/gitlab/org.md
Comment (elaine-mattos) [reply docs/integrations/gitlab/org.md]: Hi @awanlin !

I changed this line. Could you please check? :)

File: docs/integrations/gitlab/discovery.md
Comment (elaine-mattos) [reply docs/integrations/gitlab/discovery.md]: Changed this line as well!

**Comment by elaine-mattos** [comment]
Hi @vinzscam, @freben , @awanlin ,

is there anything else I can do to get this one merged? 
Please let me know if there is anything missing! :)

File: docs/integrations/gitlab/org.md
Comment (freben) [inline docs/integrations/gitlab/org.md:30]: really both?

File: plugins/catalog-backend-module-gitlab/package.json
Comment (freben) [inline plugins/catalog-backend-module-gitlab/package.json:16]:
```suggestion
    "@gitbeaker/rest": "^39.25.0",
```

if that's okay, to align with the other deps in the repo - this should clean up the yarn.lock diff

File: plugins/catalog-backend-module-gitlab/package.json
Comment (freben) [inline plugins/catalog-backend-module-gitlab/package.json:75]: not sure what happened with the field order sorting here :) version is usually way up at the top etc, but we can sort (hah!) that out some other time

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabDiscoveryEntityProvider.ts
Comment (freben) [inline plugins/catalog-backend-module-gitlab/src/providers/GitlabDiscoveryEntityProvider.ts:56]: remove

**Review by freben — APPROVED** [review]
OK, phew! I left some comments that can be deferred to a follow-up PR some time down the line. Do check them out. But for now, :shipit:

**Pull request merged**
